parsley/shared/src/main/scala/parsley/internal/deepembedding/backend/AlternativeEmbedding.scala

```diff
@@ -197,53 +221,74 @@ private [backend] object Choice {
         case alt::Nil => alt.codeGen(producesResults)
         case alt::alts_ => codeGenAlt(alt, suspend(codeGenAlternatives[M, R](alts_, producesResults)), producesResults)
     }
-    // TODO: Refactor
-    @tailrec private def foldTablified(tablified: List[(StrictParsley[_], (Char, Iterable[ExpectItem], Int, Boolean))], // scalastyle:ignore parameter.number
-                                       labelGen: CodeGenState,
-                                       roots: mutable.Map[Char, mutable.ListBuffer[StrictParsley[_]]],
-                                       backtracking: mutable.Map[Char, Boolean],
-                                       leads: mutable.ListBuffer[Char],
-                                       labels: mutable.ListBuffer[Int],
-                                       size: Int,
-                                       expecteds: List[ExpectItem],
-                                       // build in reverse!
-                                       expectedss: List[Iterable[ExpectItem]]):
-        (List[List[StrictParsley[_]]], List[Char], List[Int], Int, Iterable[ExpectItem], List[(Iterable[ExpectItem], Boolean)]) = tablified match {
-        case (root, (c, expected, _size, backtracks))::tablified_ =>
-            if (roots.contains(c)) {
-                roots(c) += root
-                backtracking(c) = backtracking(c) && backtracks
-                foldTablified(tablified_, labelGen, roots, backtracking, leads, labels,
-                              Math.max(size, _size), expected ++: expecteds, expectedss)
+
+    @tailrec private def foldTablableChars(tablified: List[TablableChar],
+                                           labelGen: CodeGenState,
+                                           roots: mutable.Map[Char, (Int, mutable.ListBuffer[StrictParsley[_]])],
+                                           map: mutable.Map[Char, (Int, Iterable[ExpectItem], Boolean)],
+                                           leads: List[Char],
+                                           size: Int,
+                                           expecteds: List[ExpectItem]):
+        (List[(Int, List[StrictParsley[_]])], mutable.Map[Char, (Int, Iterable[ExpectItem], Boolean)], Int, List[ExpectItem]) // Roots, map, size, expecteds
+            = tablified match {
+                case TablableChar(root, TablableCharDesc(c, expected, _size, backtracks)) :: tablified_ =>
+                    if (roots.contains(c)) {
+                        roots(c)._2 += root
+                        map(c) = (map(c)._1, map(c)._2, map(c)._3 && backtracks)
```

Comment:
use a `val (x, y, z) = map(c)` (with good names), this triple map lookup is wasteful